test/test_common/global.h

```diff
@@ -0,0 +1,120 @@
+#pragma once
+
+#include "common/common/lock_guard.h"
+#include "common/common/thread.h"
+
+#include "absl/container/flat_hash_map.h"
+
+namespace Envoy {
+namespace Test {
+
+/**
+ * Helper class for managing Global<Type>s.
+ *
+ * This class is instantiated as a process-scoped singleton. It manages a map
+ * from type-name to weak_ptr<GlobalHelper::Singleton>. That map accumulates
+ * over a process lifetime and never shrinks. However, the weak_ptr will
+ * generally be cleared after each test, as all shared_ptr references held in
+ * Global<Type> instances are destroyed. This way, each unit-test gets a fresh
+ * start.
+ */
+class Globals {
+  using MakeObjectFn = std::function<void*()>;
```

Comment:
Optionally might be able to base this on https://github.com/envoyproxy/envoy/blob/master/source/common/common/c_smart_ptr.h#L9.

Reply:
I had another idea; I just added a templatized derived class which handles deletion in a type-safe way (though we do still need to use static_cast to access the pointer and ref from Singleton).

ptal.